synapse/storage/databases/main/events_worker.py

```diff
@@ -736,35 +736,111 @@ async def get_stripped_room_state_from_event_context(
+            except BaseException as e:
+                exc = e
+                raise
+            finally:
+                should_restart = False
+                event_fetches_to_fail = []
+                with self._event_fetch_lock:
+                    self._event_fetch_ongoing -= 1
+                    event_fetch_ongoing_gauge.set(self._event_fetch_ongoing)
+
+                    # There may still be work remaining in `_event_fetch_list` if we
+                    # failed, or it was added in between us deciding to exit and
+                    # decrementing `_event_fetch_ongoing`.
+                    if self._event_fetch_list:
+                        if exc is None:
+                            # We decided to exit, but then some more work was added
+                            # before `_event_fetch_ongoing` was decremented.
+                            # If a new event fetch thread was not started, we should
+                            # restart ourselves since the remaining event fetch threads
+                            # may take a while to get around to the new work.
+                            # Unfortunately it is not possible to tell whether a new
+                            # event fetch thread was started, so we restart
+                            # unconditionally. If we are unlucky, we will end up with
+                            # extra idle threads holding database connections for up to
+                            # `EVENT_QUEUE_ITERATIONS * EVENT_QUEUE_TIMEOUT_S` seconds.
+                            should_restart = True
+                        elif isinstance(exc, Exception):
+                            if self._event_fetch_ongoing == 0:
+                                # We were the last remaining fetcher and failed.
+                                # Fail any outstanding fetches since no one else will
+                                # handle them.
+                                event_fetches_to_fail = self._event_fetch_list
+                                self._event_fetch_list = []
+                            else:
+                                # We weren't the last remaining fetcher, so another
+                                # fetcher will pick up the work. This will either happen
+                                # after their existing work, however long that takes,
+                                # or after at most `EVENT_QUEUE_TIMEOUT_S` seconds if
+                                # they are idle.
+                                pass
+                        else:
+                            # The exception is a `SystemExit`, `KeyboardInterrupt` or
+                            # `GeneratorExit`. Don't try to do anything clever here.
+                            pass
+
+                if should_restart:
+                    # We exited cleanly but noticed more work.
+                    self._maybe_start_fetch_thread()
+
+                if exc is not None and event_fetches_to_fail:
```

Comment:
the alignment is hard to see, but is this outside the `finally` block? It's unreachable if so - the exception will be raised at line 759

Reply:
It's inside the `finally` block still